netaddr.go

```diff
@@ -173,12 +149,14 @@ func MustParseIP(s string) IP {
 	return ip
 }
 
-// parseIPv4 parses s as an IPv4 address (in form "192.168.0.1") and
-// reports whether it parsed as such. This is a fast path for ParseIP
-// for now. It ultimately falls back to the standard library's
-// net.ParseIP at the moment, so returning ok=false from here just
-// means it'll try it a slower way.
-func parseIPv4(s string) (ip IP, ok bool) {
+func parseErr(ip, msg string, args ...interface{}) error {
```

Comment:
Call this `parseIPErr`, because we have other parsers.

I saw a tweet from you about how adding this made things slower. Fanciful theory: It's pushing stuff out of the instruction cache. Does it help if you use a bespoke error type and do the formatting work later?

```go
type ipParseError struct {
  input string
  msg string
  args ...interface{}
}

func parseIPErr(ip, msg string args ...interface{} error {
  return &ipParseError{input, msg, args}
}

func (e *ipParseError) Error() string {
  args := append([]interface{}{e.input}, e.args...)
  return fmt.Sprintf("ParseIP(%q): "+e.msg, args...)
}
```

Reply:
A nit: anything that returns a formatted string/error should be named as such, maybe "parseIPErrorf" or just "ipErrorf"?

Reply:
Another possibility is it makes the layout pass change its mind about which branch is unlikely, making the branch predictor unhappy. (The internet claims that static branch prediction isn't a thing and doesn't matter. It is empirically wrong about that.)

Reply:
Merging this PR before I dig deeper into the error things (other than the name change, which I've done). The PR is already quite large and even with the mild regression introduced by error handling, it's much better than what's at HEAD right now.